Documentation update on tickless mode porting guide (#2)
rajkan01/mbed-os-5-docs: docs/porting/target/tickless/tickless.md

# Tickless

## Tickless mode

> Mbed OS has SysTick timer which generates tick in a periodic interval. The interrupt generated on every tick to reschedule the threads. The periodic tick configures as 1ms for Mbed OS.

Reviewer: ```suggestion
Mbed OS has a SysTick timer which generates periodic ticks. An interrupt is generated on every tick to reschedule the threads. The SysTick period is 1 ms for Mbed OS.
```

![](./resources/Tick.png)

> Tickless mode is an optimization mechanism to suspend the SysTick timer. In particular, this will put the system to uninterrupted sleep to achieve power saving when idle for multiple ticks. Target in tickless mode normally disables the SysTick, sets up wake-up timers and enters sleep mode when idle. When SysTick timer expires on its own or from an external event like interrupt, then it exits sleep mode and re-enables the SysTick.

Reviewer: ```suggestion
Tickless mode is an optimization mechanism to suspend the SysTick timer when idle. In particular, the system will enter deep sleep to achieve power saving when idle for multiple ticks. An idle target in tickless mode disables the SysTick timer, sets up a low power wake-up timer and enters deep sleep mode. When the wake-up timer fires, or an external interrupt occurs, the target exits sleep mode and re-enables the SysTick timer.
```

Reviewer 2: @gpsimenos Although the original text needs editing, I liked that in the introduction to tickless there was no reference to deep sleep. Tickless means no SysTick and therfore sleep when the system is idle. Subsequently we can introduce usage of low power timer to allow deep sleep mode.

Reviewer 2: @rajkan01 The original text `Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can achieve power savings by entering uninterrupted sleep.` was clearer.

> [Mbed OS sleep modes](../../../apis/power-management-sleep.html):

Reviewer 2: An introduction to the Mbed OS sleep modes is missing.
For example, you could say that a further power optimisation is achieved by entering deep sleep when the system is idle.

- Sleep mode: Both low and high-frequency clocks are active.

> - Deep sleep mode: Only the low-frequency clock is active.

Reviewer: ```suggestion
- Deep sleep mode: Only the low-frequency, low-power clock is active.
```

> To enable tickless mode in Mbed OS, your target has to meet following requirements:

Reviewer 2: Before this section you should explain the issue with using two clocks, system timer and low power timer. This causes drift. To avoid it, Mbed OS overrides SysTick with the low power timer.

Reviewer 2: You could then say that it is also possible to enable tickless with the microsecond ticker only. This avoids the drift issue (one source clock only) but prevents deep sleep.

- Support for sleep HAL API
- Support for low power or microsecond ticker HAL API

> If the target supports both low power and microsecond ticker, then low power ticker is selected with tickless mode.

Reviewer: ```suggestion
If the target supports both low power and microsecond ticker, then low power ticker is selected with tickless mode. The greatest power savings are achieved when a target has a low power ticker.
```

> For instance, the system runs only with two threads thread 1 and thread 2. During the first 2 tick interrupts thread 1 and thread 2 schedules alternatively. When thread 1 entered into a wait state, the scheduler preempts and schedules thread 2 for execution. After some time, thread 2 enters into a wait state and get preempted. Now scheduler schedules idle thread to run as two threads are in the wait state. This idle thread put the system to sleep. After multiple SysTick periodic interrupts, thread 2 is rescheduled for execution. All previous Systick wake-ups waste power.

Reviewer: ```suggestion
For instance, consider the system is running in non-tickless mode with two threads, thread 1 and thread 2. During the first 2 SysTick interrupts thread 1 and thread 2 are scheduled alternatively. When thread 1 enters a wait state, the scheduler preempts and schedules thread 2 for execution. After some time, thread 2 enters a wait state and gets preempted. Now the scheduler schedules the idle thread to run as both threads are in the wait state. This idle thread puts the system to sleep. However, the SysTick timer is still running, regularly waking up the system even though no threads are active. Those unnecessary Systick wake-ups waste power.
```

> For instance, the system runs only with two threads thread 1 and thread 2. During the first 2 tick interrupts thread 1 and thread 2 schedules alternatively. When thread 1 entered into a wait state, the scheduler preempts and schedules thread 2 for execution. After some time, thread 2 enters into a wait state and get preempted. Now scheduler schedules idle thread to run as two threads are in the wait state. This idle thread put the system to sleep. After multiple SysTick periodic interrupts, thread 2 is rescheduled for execution. All previous Systick wake-ups waste power.

Reviewer 2: This section with the two time diagrams should be moved earlier in the page.

![](./resources/Normal_Tick.png)

> But with tickless mode, it suspends the SysTick periodic wake up as no threads are ready to run for quite some time which inturn saves the power.

Reviewer: ```suggestion
Instead, when tickless mode is enabled, it suspends the SysTick periodic wake up as no threads are ready to run. That saves significant power.
```

![](./resources/Tickless.png)

### Enable tickless mode support in full profile

Find your target in `mbed-os/targets/targets.json` and update its configuration as described below.

1. Add `MBED_TICKLESS` macro into the `macros` option as follows:

    ```json
	"macros": [
        "MBED_TICKLESS"
    ]    
    ```

> 1. If your target does not have a low power ticker or you prefer to use microsecond ticker instead of low power ticker for tickless mode, then ensure that tickless is enabled using the microsecond ticker:

Reviewer 2: Why would you prefer to use the microsecond ticker? Can you think of some use cases?

    ```json
    "overrides": {
        "tickless-from-us-ticker": true
    }
    ```

> Targets supporting tickless mode override the default SysTick mechanism and use the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) if "tickless-from-us-ticker" target configuration is set. This change is necessary to avoid drift connected with using two different timers to measure time. Users must not change the ticker interrupt handler when tickless mode is in use.

Reviewer 2: This paragraph should be moved somewhere else. See earlier comment.

> Targets supporting tickless mode override the default SysTick mechanism and use the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) if "tickless-from-us-ticker" target configuration is set. This change is necessary to avoid drift connected with using two different timers to measure time. Users must not change the ticker interrupt handler when tickless mode is in use.

Reviewer 2: Can you provide more explanation to `Users must not change the ticker interrupt handler when tickless mode is in use.`?

### Enable tickless mode support in bare metal profile 

> Tickless mode is enabled by default for bare metal profile without adding `MBED_TICKLESS` target configuration. As mentioned earlier target prefer to use microsecond ticker for tickless mode or does not have low power ticker then add the below target configuration in `targets.json`.

Reviewer: ```suggestion
Tickless mode is enabled by default for bare metal profile without adding `MBED_TICKLESS` target configuration. As mentioned earlier, if you prefer to use microsecond ticker for tickless mode, or the target does not have a low power ticker, then add the below target configuration in `targets.json`.
```

Reviewer 3: As described above, Tickless means "pausing" the SysTick, sleep for a duration, then resume the SysTick. But if the 1ms-SysTick is for thread scheduling, is it still enabled on bare-metal? How do we say "Tickless mode is enabled by default for bare metal profile"?

Author: You are correct. SysTick is never used in bare metal.
But timed sleeping RTOS API calls like _Semaphore::try_acquire_for_ or _ThisThread::sleep_until_, it directly requests a wake-up from the SysTimer and calls sleep.

Reviewer 2: @LDong-Arm is exactly rigth. It is not right to say that "tickless is enabled by default for bare metal" but rather that tickless is a RTOS optimisation, it has no meaning in bare metal.

Reviewer 2: @rajkan01 Look at the documentation for the Mbed OS Timer: OsTimer

This provides the marshalling point for a system global SysTimer, which is used to provide:
- timed sleeps (for default idle hook in RTOS tickless mode, or non-RTOS sleeps)
- regular ticks for RTOS
- absolute system timing (directly for non-RTOS, or indirectly via RTOS tick count)

Bare metal uses SysTimer to implement timed sleeps. This is a different matter.

```
"overrides": {
    "tickless-from-us-ticker": true
}
```

### Testing

> There are no dedicated tests validating tickless mode. Running all Mbed OS tests suits, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.

Reviewer: ```suggestion
There are no dedicated tests for validating tickless mode. Running all Mbed OS tests suits, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.
```

> There are no dedicated tests validating tickless mode. Running all Mbed OS tests suits, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.

Reviewer 2: This paragraph might need rewriting. Currently we do not have dedicated tests but we could (for example we should ensure that there is no clock drift. We could also ensure that higher power savings are achieved). Are we sure that the current HAL tests provide enough coverage?

Reviewer 2: What is missing at the end is some notes on current issues/challenges with tickless. @gpsimenos can you track why your changes to enable tickless were reverted?

> Mbed OS has SysTick timer which generates tick in a periodic interval. The interrupt generated on every tick to reschedule the threads. The periodic tick configures as 1ms for Mbed OS.
> 
> ![](./resources/Tick.png)
> 
> Tickless mode is an optimization mechanism to suspend the SysTick timer. In particular, this will put the system to uninterrupted sleep to achieve power saving when idle for multiple ticks. Target in tickless mode normally disables the SysTick, sets up wake-up timers and enters sleep mode when idle. When SysTick timer expires on its own or from an external event like interrupt, then it exits sleep mode and re-enables the SysTick.
> 
> [Mbed OS sleep modes](../../../apis/power-management-sleep.html):
> 
> - Sleep mode: Both low and high-frequency clocks are active.
> - Deep sleep mode: Only the low-frequency clock is active.
> 
> To enable tickless mode in Mbed OS, your target has to meet following requirements:
> 
> - Support for sleep HAL API
> - Support for low power or microsecond ticker HAL API
> 
> If the target supports both low power and microsecond ticker, then low power ticker is selected with tickless mode.
> 
> For instance, the system runs only with two threads thread 1 and thread 2. During the first 2 tick interrupts thread 1 and thread 2 schedules alternatively. When thread 1 entered into a wait state, the scheduler preempts and schedules thread 2 for execution. After some time, thread 2 enters into a wait state and get preempted. Now scheduler schedules idle thread to run as two threads are in the wait state. This idle thread put the system to sleep. After multiple SysTick periodic interrupts, thread 2 is rescheduled for execution. All previous Systick wake-ups waste power.
> 
> ![](./resources/Normal_Tick.png)
> 
> But with tickless mode, it suspends the SysTick periodic wake up as no threads are ready to run for quite some time which inturn saves the power.
> 
> ![](./resources/Tickless.png)
> 
> ### Enable tickless mode support in full profile
> 
> Find your target in `mbed-os/targets/targets.json` and update its configuration as described below.
> 
> 1. Add `MBED_TICKLESS` macro into the `macros` option as follows:
> 
>     ```json
> 	"macros": [
>         "MBED_TICKLESS"
>     ]    
>     ```
> 1. If your target does not have a low power ticker or you prefer to use microsecond ticker instead of low power ticker for tickless mode, then ensure that tickless is enabled using the microsecond ticker:
> 
>     ```json
>     "overrides": {
>         "tickless-from-us-ticker": true
>     }
>     ```
> 
> Targets supporting tickless mode override the default SysTick mechanism and use the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) if "tickless-from-us-ticker" target configuration is set. This change is necessary to avoid drift connected with using two different timers to measure time. Users must not change the ticker interrupt handler when tickless mode is in use.
> 
> ### Enable tickless mode support in bare metal profile 
> 
> Tickless mode is enabled by default for bare metal profile without adding `MBED_TICKLESS` target configuration. As mentioned earlier target prefer to use microsecond ticker for tickless mode or does not have low power ticker then add the below target configuration in `targets.json`.
> 
> ```
> "overrides": {
>     "tickless-from-us-ticker": true
> }
> ```
> 
> ### Testing
> 
> There are no dedicated tests validating tickless mode. Running all Mbed OS tests suits, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.

Reviewer 4: ```suggestion
Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can achieve power savings by entering uninterrupted sleep. 

# Reminder on scheduling & sleep modes in Mbed OS

Mbed OS uses the SysTick timer at period of 1ms to process threads' scheduling.
For instance, a system running two threads would see its timing diagram look like :

![](./resources/Normal_Tick.png)
<!-- Please fix the names on the diagram. IIRC `wait` is now either deprecated or removed. It should show ThisThread::sleep_for() or thread_sleep_for https://os.mbed.com/docs/mbed-os/v5.15/apis/thisthread.html -->

Note how the device never enters deepsleep and wastes cycles in Systick while all thread are asleep.

# Tickless mode

To support tickless mode in Mbed OS, your target needs to meet two requirements:
- Support for Sleep HAL API
- Support for Low Power Ticker HAL API
To enable tickless mode support in Mbed OS, you need to add the `MBED_TICKLESS` macro in the macros option of the target's section in the `targets.json` file.

When the tickless mode is enabled, the device suspends the kernel effectively disabling the systick and sets up the [low power ticker] through [mbed::internal::OsTimer].

If the target does not feature a low power ticker or has an excessively long wake-up time, you must also set the `tickless-from-us-ticker` to true in the target's section in the `targets.json` file. In such the device will only enter sleep rather than deepsleep.

*Note that using the tickless mode prevents the user from using the associated timer.*

The expected scheduling for the same use-case as previously described should look like :

![](./resources/Tickless.png)
<!-- Please fix the names on the diagram. IIRC `wait` is now either deprecated or removed. It should show ThisThread::sleep_for() or thread_sleep_for https://os.mbed.com/docs/mbed-os/v5.15/apis/thisthread.html -->

# Testing

There are no dedicated tests validating tickless mode. Running all Mbed OS tests suites, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.

# References

You can find more details on scheduling in mbed in the section : <!-- use appropriate link here -->
You can find more details on sleep modes in mbed in the section :<!-- power-management-sleep.html use appropriate link here -->
```
